crates/synthesis/src/multi_controlled/mcx.rs

```diff
@@ -0,0 +1,430 @@
+    /// The operations are added in-place.
+    fn compose(&mut self, other: &Self, qargs_map: &[Qubit], cargs_map: &[Clbit]) -> PyResult<()> {
+        for inst in other.data() {
+            let remapped_qubits: Vec<Qubit> = other
+                .get_qargs(inst.qubits)
+                .iter()
+                .map(|q| qargs_map[q.index()])
+                .collect();
+            let remapped_clbits: Vec<Clbit> = other
+                .get_cargs(inst.clbits)
+                .iter()
+                .map(|c| cargs_map[c.index()])
+                .collect();
+
+            self.push_packed_operation(
+                inst.op.clone(),
+                inst.params_view(),
+                &remapped_qubits,
+                &remapped_clbits,
+            );
+        }
+
+        self.add_global_phase(other.global_phase())?;
+        Ok(())
+    }
+
+    /// Construct the inverse circuit
+    fn inverse(&self) -> PyResult<CircuitData> {
+        let inverse_global_phase =
+            Python::with_gil(|py| -> Param { multiply_param(self.global_phase(), -1.0, py) });
+
+        let mut inverse_circuit = CircuitData::clone_empty_like(self, None)?;
+        inverse_circuit.set_global_phase(inverse_global_phase)?;
+
+        let data = self.data();
+
+        for i in 0..data.len() {
+            let inst = &data[data.len() - 1 - i];
+
+            let inverse_inst: Option<(StandardGate, SmallVec<[Param; 3]>)> = match &inst.op.view() {
+                OperationRef::StandardGate(gate) => gate.inverse(inst.params_view()),
+                _ => None,
+            };
+
+            if inverse_inst.is_none() {
+                return Err(CircuitError::new_err(format!(
+                    "The circuit cannot be inverted: {} is not a standard gate.",
+                    inst.op.name()
+                )));
+            }
+
+            let (inverse_op, inverse_op_params) = inverse_inst.unwrap();
+
+            inverse_circuit.push_packed_operation(
+                inverse_op.into(),
+                &inverse_op_params,
+                self.get_qargs(inst.qubits),
+                self.get_cargs(inst.clbits),
+            );
+        }
+        Ok(inverse_circuit)
+    }
+}
+
+/// Efficient synthesis for 4-controlled X-gate.
+#[pyfunction]
+pub fn c4x() -> PyResult<CircuitData> {
+    let mut circuit = CircuitData::with_capacity(5, 0, 0, Param::Float(0.0))?;
+    circuit.h(4);
+    circuit.cp(PI2, 3, 4);
+    circuit.h(4);
+    circuit.compose(&rc3x()?, &[Qubit(0), Qubit(1), Qubit(2), Qubit(3)], &[])?;
+    circuit.h(4);
+    circuit.cp(-PI2, 3, 4);
+    circuit.h(4);
+    circuit.compose(
+        &rc3x()?.inverse()?,
+        &[Qubit(0), Qubit(1), Qubit(2), Qubit(3)],
+        &[],
+    )?;
+    circuit.compose(&c3sx()?, &[Qubit(0), Qubit(1), Qubit(2), Qubit(4)], &[])?;
+    Ok(circuit)
+}
+
+/// A block in the `action part`, see Iten et al.
+fn action_gadget() -> PyResult<CircuitData> {
```

Comment:
Could this be moved into `add_action_gadget`, since this function is not used as standalone? That would also skip the additional creation of a `CircuitData`